Corrupted attester tmp directory

Hello,

I have currently integrated attester plugin into 1A UI build process to facilitate UI test-driven development.
However, I notice that during some random circumstances the extracted attester directory is corrupted on the server and has caused all pull requests being rejected.

Here's the log:
[ERROR] Failed to execute goal com.ariatemplates.attester:attester-maven-plugin:1.5.0:test (default) on project aodb_ui: Execution default of goal com.ariatemplates.attester:attester-maven-plugin:1.5.0:test failed: Failed to find or extract attester: Could not find file: /tmp/com.ariatemplates.attester-attester-1.5.0-project/attester/bin/attester.js -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.ariatemplates.attester:attester-maven-plugin:1.5.0:test (default) on project aodb_ui: Execution default of goal com.ariatemplates.attester:attester-maven-plugin:1.5.0:test failed: Failed to find or extract attester
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:224)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default of goal com.ariatemplates.attester:attester-maven-plugin:1.5.0:test failed: Failed to find or extract attester
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:143)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	... 19 more
Caused by: java.lang.RuntimeException: Failed to find or extract attester
	at com.ariatemplates.attester.maven.RunAttester.extractDependency(RunAttester.java:277)
	at com.ariatemplates.attester.maven.RunAttester.extractAttester(RunAttester.java:283)
	at com.ariatemplates.attester.maven.RunAttester.execute(RunAttester.java:289)
	at com.ariatemplates.attester.maven.RunTests.execute(RunTests.java:105)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
	... 20 more
Caused by: java.io.FileNotFoundException: Could not find file: /tmp/com.ariatemplates.attester-attester-1.5.0-project/attester/bin/attester.js
	at com.ariatemplates.attester.maven.RunAttester.extractDependency(RunAttester.java:274)
	... 24 more

Quick investigation reveals that this issue may be linked to parallel builds on the server. I can simply remove the tmp directory but this won't completely solve the root cause.

I have 2 proposals to overcome the issue:
* extract attester framework all the time by prefixing directory with timestamp or random values and delete it after test completion
or
* enhance the logic to validate attester directory is available and not corrupted else override it

Could you please advise if you can implement any of these proposals?
Thanks & Best regards,
Imam

--- a/src/main/java/com/ariatemplates/attester/maven/ArtifactExtractor.java
+++ b/src/main/java/com/ariatemplates/attester/maven/ArtifactExtractor.java
@@ -23,7 +23,6 @@ import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.OutputStream;
 import java.util.Map;
-import java.util.regex.Pattern;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipInputStream;
 
@@ -67,6 +66,15 @@ public class ArtifactExtractor extends AbstractMojo {
      */
     protected MavenProject project;
 
+    /**
+     * Parent directory in which directories containing the extracted artifacts
+     * will be created. If this parameter is not set, the temporary directory
+     * defined in the <code>java.io.tmpdir</code> system property is used.
+     *
+     * @parameter
+     */
+    public File outputParentDirectory;
+
     public void execute() throws MojoExecutionException, MojoFailureException {
         try {
             Map<?, ?> context = getPluginContext();
@@ -88,12 +96,18 @@ public class ArtifactExtractor extends AbstractMojo {
     }
 
     public static File getOutputDirectory(Artifact artifact) {
-        File tempDirectory = new File(System.getProperty("java.io.tmpdir"));
+        return getOutputDirectory(artifact, null);
+    }
+
+    public static File getOutputDirectory(Artifact artifact, File parentDirectory) {
+        if (parentDirectory == null) {
+            parentDirectory = new File(System.getProperty("java.io.tmpdir"));
+        }
         String name = artifact.getGroupId() + "-" + artifact.getArtifactId() + "-" + artifact.getVersion() + "-" + artifact.getClassifier();
         if (artifact.isSnapshot()) {
             name += "-" + artifact.getFile().lastModified();
         }
-        File res = new File(tempDirectory, name);
+        File res = new File(parentDirectory, name);
         return res;
     }
 
@@ -102,7 +116,7 @@ public class ArtifactExtractor extends AbstractMojo {
             dependency.getType(), dependency.getClassifier(), new DefaultArtifactHandler("zip"));
         artifact = localRepository.find(artifact);
         File zipFile = artifact.getFile();
-        File outputDirectory = getOutputDirectory(artifact);
+        File outputDirectory = getOutputDirectory(artifact, outputParentDirectory);
         if (outputDirectory.exists()) {
             // if the directory already exists, don't touch it (by the way, it
             // may currently be in use by some other process)

--- a/src/main/java/com/ariatemplates/attester/maven/RunAttester.java
+++ b/src/main/java/com/ariatemplates/attester/maven/RunAttester.java
@@ -230,6 +230,18 @@ public abstract class RunAttester extends RunNode {
     public File attesterPath;
 
     /**
+     * Parent directory in which the directory containing the extracted attester files
+     * will be created, if it does not already exist.
+     * Files are extracted from the following maven artifact:
+     * <code>com.ariatemplates.attester:attester:zip:project</code>
+     *
+     * This parameter is only used if attesterPath is not defined.
+     *
+     * @parameter expression="${project.build.directory}"
+     */
+    public File attesterExtractionParentDirectory;
+
+    /**
      * Port for the internal web server. (Passed through <code>--port</code> to
      * <a href="https://github.com/ariatemplates/attester#usage" >attester</a>)
      *
@@ -266,6 +278,7 @@ public abstract class RunAttester extends RunNode {
                 res = new File(property, pathAfterProperty);
             } else {
                 ArtifactExtractor extractor = new ArtifactExtractor();
+                extractor.outputParentDirectory = attesterExtractionParentDirectory;
                 extractor.setLog(this.getLog());
                 String outputDirectory = extractor.inplaceExtractDependency(session.getLocalRepository(), dependency);
                 res = new File(outputDirectory, pathAfterDependency);

